python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py

```diff
@@ -0,0 +1,90 @@
+***************************************************************************
+*                                                                         *
+*   This program is free software; you can redistribute it and/or modify  *
+*   it under the terms of the GNU General Public License as published by  *
+*   the Free Software Foundation; either version 2 of the License, or     *
+*   (at your option) any later version.                                   *
+*                                                                         *
+***************************************************************************
+"""
+
+__author__ = 'Anita Graser'
+__date__ = 'May 2018'
+__copyright__ = '(C) 2018, Anita Graser'
+
+# This will get replaced with a git SHA1 when you do a git archive
+
+__revision__ = '$Format:%H$'
+
+from qgis.core import (Qgis, QgsProcessingException, QgsProcessingParameterString, QgsApplication, QgsVectorLayer, QgsProject, QgsProcessingParameterFeatureSink, QgsProcessing, QgsFeatureRequest, QgsFeature, QgsFeatureSink)
+from processing.algs.qgis.QgisAlgorithm import QgisAlgorithm
+from processing.tools import postgis
+
+
+class PostGISExecuteAndLoadSQL(QgisAlgorithm):
+
+    DATABASE = 'DATABASE'
+    SQL = 'SQL'
+    OUTPUT = 'OUTPUT'
+
+    def group(self):
+        return self.tr('Database')
+
+    def groupId(self):
+        return 'database'
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        db_param = QgsProcessingParameterString(
+            self.DATABASE,
+            self.tr('Database (connection name)'))
+        db_param.setMetadata({
+            'widget_wrapper': {
+                'class': 'processing.gui.wrappers_postgis.ConnectionWidgetWrapper'}})
+        self.addParameter(db_param)
+        self.addParameter(QgsProcessingParameterString(self.SQL, self.tr('SQL query (must return unique id and geom field)'), multiLine=True))
+        self.addParameter(QgsProcessingParameterFeatureSink(
+            self.OUTPUT,
+            self.tr("Output layer"),
+            QgsProcessing.TypeVectorAnyGeometry))
+
+    def name(self):
+        return 'postgisexecuteandloadsql'
+
+    def displayName(self):
+        return self.tr('PostGIS execute and load SQL')
+
+    def processAlgorithm(self, parameters, context, feedback):
+        connection = self.parameterAsString(parameters, self.DATABASE, context)
+        #db = postgis.GeoDB.from_name(connection)
+        uri = postgis.uri_from_name(connection)
+        sql = self.parameterAsString(parameters, self.SQL, context).replace('\n', ' ')
+        QgsApplication.messageLog().logMessage(str(sql), level=Qgis.Info)
+        uri.setDataSource("","("+sql+")", "geom", "","id")
+        vlayer = QgsVectorLayer(uri.uri(), "layername", "postgres")
+        QgsApplication.messageLog().logMessage("Valid layer: "+str(vlayer.isValid()), level=Qgis.Info)
+        #QgsProject.instance().addMapLayer(vlayer)
```

Comment:
Was this the original intention here? If so, you could do this by:

    def processAlgorithm(self, parameters, context, feedback):
         # ...everything required to build up the uri...
         layer = QgsProcessingUtils.mapLayerFromString( uri.uri(), context )
         if layer is None:
              raise QgsProcessingException(self.tr("some helpful error message"))

         context.addLayerToLoadOnCompletion( layer.id(), QgsProcessingContext.LayerDetails( 'some nice name', context.project(), self.OUTPUT ) )

         return { self.OUTPUT: layer.id() }
        
If you go this way, you'd need to change OUTPUT from a QgsProcessingParameterFeatureSink to a QgsProcessingOutputVectorLayer output.

Reply:
Yes, that was the original intention. Unfortunately, I'm having no success with QgsProcessingUtils.mapLayerFromString( uri.uri(), context ). The resulting layer is always None even though the uri works with the previous code.

Reply:
Not sure how to debug mapLayerFromString's silent refusal to return a layer.

Reply:
Thanks, that worked!